### PR TITLE
[NOMERGE] Merge window creation, attempt 2

### LINKED ISSUE #1619: Make API fully backend-agnostic
This issue tracks progress in making API usage uniform accross backends.

As it currently stands, usage of gfx is still specific to the backend chosen, forcing cross-backend code to be feature gated – _see [examples/hal/quad](https://github.com/gfx-rs/gfx/blob/master/examples/hal/quad/main.rs)_:

---
### To-do:
- [x] Different entry point name for Metal:
```rust
// MSL doesn't allow `main` entry point name.
//TODO: just use a different name consistently in all backends
#[cfg(feature = "metal")]
const ENTRY_NAME: &str = "main0";
#[cfg(not(feature = "metal"))]
const ENTRY_NAME: &str = "main";
```
```rust
        //TODO: remove the type annotations when we have support for
        // inidirect argument buffers in SPIRV-Cross
        #[cfg(any(feature = "vulkan", feature = "dx12", feature = "metal", feature = "gl"))]
        let (vs_entry, fs_entry) = (
            pso::EntryPoint::<back::Backend> { entry: ENTRY_NAME, module: &vs_module },
            pso::EntryPoint::<back::Backend> { entry: ENTRY_NAME, module: &fs_module },
        );

        #[cfg(all(feature = "metal", feature = "metal_argument_buffer"))]
        let (vs_entry, fs_entry) = (
            pso::EntryPoint { entry: "vs_main", module: &shader_lib },
            pso::EntryPoint { entry: "ps_main", module: &shader_lib },
);
```
- [x] Autorelease pool for Metal:
```rust
#[cfg(feature = "metal")]
    let mut autorelease_pool = unsafe { back::AutoreleasePool::new() };
```
- [ ] Different window creation for OpenGL:
```rust
   #[cfg(any(feature = "vulkan", feature = "dx12", feature = "metal"))]
    let window = wb
        .build(&events_loop)
        .unwrap();
    #[cfg(feature = "gl")]
    let window = {
        let builder = back::config_context(
            back::glutin::ContextBuilder::new(),
            ColorFormat::SELF,
            None,
        ).with_vsync(true);
        back::glutin::GlWindow::new(wb, builder, &events_loop).unwrap()
};
```
- [ ] Different backend instantiation for OpenGL:
```rust
    #[cfg(any(feature = "vulkan", feature = "dx12", feature = "metal"))]
    let (_instance, mut adapters, mut surface) = {
        let instance = back::Instance::create("gfx-rs quad", 1);
        let surface = instance.create_surface(&window);
        let adapters = instance.enumerate_adapters();
        (instance, adapters, surface)
    };
    #[cfg(feature = "gl")]
    let (mut adapters, mut surface) = {
        let surface = back::Surface::from_window(window);
        let adapters = surface.enumerate_adapters();
(
```
- [x] #71 Non uniform handling of shader module creation:
```rust
// Setup renderpass and pipeline
    #[cfg(any(feature = "vulkan", feature = "dx12", feature = "metal"))]
    let vs_module = device
        .create_shader_module(include_bytes!("data/vert.spv"))
        .unwrap();
    #[cfg(any(feature = "vulkan", feature = "dx12", feature = "metal"))]
    let fs_module = device
        .create_shader_module(include_bytes!("data/frag.spv"))
        .unwrap();

    #[cfg(all(feature = "metal", feature = "metal_argument_buffer"))]
    let shader_lib = device.create_shader_library_from_source(
            include_str!("shader/quad_indirect.metal"),
            back::LanguageVersion::new(2, 0),
        ).expect("Error on creating shader lib");

    #[cfg(feature = "gl")]
    let vs_module = device
        .create_shader_module_from_source(
            include_bytes!("shader/quad_150.glslv"),
            pso::Stage::Vertex,
        ).unwrap();
    #[cfg(feature = "gl")]
    let fs_module = device
        .create_shader_module_from_source(
            include_bytes!("shader/quad_150.glslf"),
            pso::Stage::Fragment,
).unwrap();
```

### PR DESCRIPTION
**Pull request by ZeGentzy**
I've been sitting on this for a while, so here it is. Attempt 1: https://github.com/gfx-rs/gfx/pull/2136 

Signed-off-by: Hal Gentz <zegentzy@protonmail.com>

Fixes #1619
PR checklist:

- [x] Implement on X11

Depends on: https://github.com/tomaka/glutin/pull/1058 https://github.com/tomaka/winit/pull/627  https://github.com/tomaka/glutin/pull/1029

### CODE REVIEW

File: src/backend/gl/src/queue.rs [hunk 1]
```diff
@@ -730,12 +730,58 @@ impl hal::queue::RawCommandQueue<Backend> for CommandQueue {
...
+            );
+
+            if extent != swapchain.extent {
+                return Err(());
+            }
+
+            unsafe {
+                let gl = &swapchain.surface.device.share.context;
+                let fbo = swapchain.images[index as usize].fbo;
+                gl.BindFramebuffer(gl::READ_FRAMEBUFFER, fbo);
```
Comment (ZeGentzy) [inline src/backend/gl/src/queue.rs]: Won't work with OpenGL 2.1 without the `EXT_framebuffer_object` and `EXT_framebuffer_bind` extensions. Even then we would have to append `EXT` to all the relevant functions and `_EXT` to all the relevant constants.

File: src/backend/gl/src/window/glutin.rs [hunk 2]
```diff
@@ -43,60 +43,152 @@
...
+            .get_inner_size()
+            .unwrap()
+            .to_physical(self.surface.surface_context.window.get_hidpi_factor())
+        );
+
+        if extent != self.extent {
+            return Err(hal::AcquireError::OutOfDate);
+        }
+
+        use std::time::Instant;
```
Comment (ZeGentzy) [inline src/backend/gl/src/window/glutin.rs]: Respecting the timeout could be split into a separate PR, if someone would like.

File: src/backend/gl/src/window/glutin.rs [hunk 3]
```diff
@@ -43,60 +43,152 @@
...
     }
 }
 
-//TODO: if we make `Surface` a `WindowBuilder` instead of `GlWindow`,
-// we could spawn window + GL context when a swapchain is requested
-// and actually respect the swapchain configuration provided by the user.
-pub struct Surface {
-    window: Starc<glutin::GlWindow>,
+#[cfg(all(unix, not(target_os = "android")))]
+pub struct SurfacePreSwapchained {
```
Comment (ZeGentzy) [inline src/backend/gl/src/window/glutin.rs]: Might be better as an enum, as we will never have both at the same time.

File: src/backend/gl/src/window/glutin.rs [hunk 4]
```diff
@@ -118,33 +210,93 @@ impl Surface {
...
+                    .get_pixel_format()
+                    .multisampling
+                    .unwrap_or(1);
+                hal::image::Kind::D2(extent.width, extent.height, 1, samples as _)
+            }
+            _ => unimplemented!(),
+        }
     }
 
+    // TODO return proper requirments, currently doesn't matter because we just
```
Comment (ZeGentzy) [inline src/backend/gl/src/window/glutin.rs]: I personally don't think addressing this is relevant to this PR, but that's debatable.

File: src/backend/gl/src/window/glutin.rs [hunk 5]
```diff
@@ -118,33 +210,93 @@ impl Surface {
...
+                    f::Format::Bgra8Unorm,
+                ];
+                (None, extents, 2..3, formats)
+            }
+            Surface::Swapchained(surface) => {
+                // NOTE: Once swapchained, we cannot change a surface's
+                // properties. That's why we return the surface's properties,
+                // so those pesky applications don't try to change it.
+                //
+                // NOTE: What we are currently returning *might* actually be
```
Comment (ZeGentzy) [inline src/backend/gl/src/window/glutin.rs]: Note says it all.

File: src/backend/gl/src/window/glutin.rs [hunk 6]
```diff
@@ -154,55 +306,424 @@ impl Device {
...
+                    let ic = Wstarc::upgrade(ic).unwrap();
+                    let ctxt = &ic.context;
+
+                    if let Some(ref xci) = s.xci {
+                        ContextMaker::new_xlib(
+                            ctxt,
+                            &xci,
+                        ).unwrap()
+                    } else if let Some(ref wci) = s.wci {
+                        // TODO: Someone should test that this doesn't fail if
```
Comment (ZeGentzy) [inline src/backend/gl/src/window/glutin.rs]: Note says it all.

File: src/backend/gl/src/window/glutin.rs [hunk 7]
```diff
@@ -154,55 +306,424 @@ impl Device {
...
+                }
+
+                use hal::Device;
+                let kind = image::Kind::D2(config.extent.width, config.extent.height, 1, 1);
+                let uimage = surface.device.create_image(
+                    kind,
+                    1,
+                    f::Rgba8Srgb::SELF,
+                    image::Tiling::Optimal, // ignored
+                    image::Usage::SAMPLED, // FIXME: Hello reviewer! Is this the correct one? I'm trying to get a texture, not a surface, and this appears to be the way.
```
Comment (ZeGentzy) [inline src/backend/gl/src/window/glutin.rs]: Is this the correct one? I'm trying to get a texture, not a surface, and this appears to be the way.

File: src/backend/gl/src/window/glutin.rs [hunk 8]
```diff
@@ -154,55 +306,424 @@ impl Device {
...
+                    gl.GenFramebuffers(1, &mut fbo);
+                }
+
+                use hal::Device;
+                let kind = image::Kind::D2(config.extent.width, config.extent.height, 1, 1);
+                let uimage = surface.device.create_image(
+                    kind,
+                    1,
+                    f::Rgba8Srgb::SELF,
+                    image::Tiling::Optimal, // ignored
```
Comment (ZeGentzy) [inline src/backend/gl/src/window/glutin.rs]: Why is it ignored? I just copied the comment from another place.

File: src/backend/gl/src/window/glutin.rs [hunk 9]
```diff
@@ -154,55 +306,424 @@ impl Device {
...
+                    .bind_image_memory(&image_mem, 0, uimage)
+                    .unwrap();
+
+                unsafe {
+                    let gl = &surface.device.share.context;
+                    let image = match image.kind {
+                        native::ImageKind::Texture(t) => t,
+                        _ => panic!(),
+                    };
+                    gl.BindFramebuffer(gl::FRAMEBUFFER, fbo);
```
Comment (ZeGentzy) [inline src/backend/gl/src/window/glutin.rs]: Like I said above about OpenGL 2.1.

File: src/backend/gl/src/window/glutin.rs [hunk 10]
```diff
@@ -154,55 +306,424 @@ impl Device {
...
-impl hal::Instance for Surface {
-    type Backend = B;
-    fn enumerate_adapters(&self) -> Vec<hal::Adapter<B>> {
-        unsafe { self.window.make_current().unwrap() };
-        let adapter = PhysicalDevice::new_adapter(|s| self.window.get_proc_address(s) as *const _);
-        vec![adapter]
+pub struct ContextMaker;
+
+#[derive(Debug)]
+pub enum ContextCreationError {
```
Comment (ZeGentzy) [inline src/backend/gl/src/window/glutin.rs]: Could use improvements.

File: src/backend/gl/src/window/glutin.rs [hunk 11]
```diff
@@ -154,55 +306,424 @@ impl Device {
...
+        let ctxt = unsafe {
+            glutin::GlSeparatedContext::new_shared(&win, cb, &evlp)
+                .map_err(|err| ContextCreationError::GlutinCreationError(err))?
+        };
+
+        Ok((win, evlp, ctxt))
+    }
+
+    #[cfg(all(unix, not(target_os = "android")))]
+    pub fn new_wayland(
```
Comment (ZeGentzy) [inline src/backend/gl/src/window/glutin.rs]: Doesn't actually currently work on wayland. :/

File: src/backend/vulkan/Cargo.toml [hunk 12]
```diff
@@ -26,7 +26,8 @@ shared_library = { version = "0.1.9", optional = true }
 ash = "0.24.4"
 gfx-hal = { path = "../../hal", version = "0.1" }
 smallvec = "0.6"
-winit = { version = "0.17", optional = true }
+#winit = { version = "0.17", optional = true }
```
Comment (ZeGentzy) [inline src/backend/vulkan/Cargo.toml]: Should be undone when finished.

File: examples/Cargo.toml [hunk 13]
```diff
@@ -28,7 +28,8 @@ path = "compute/main.rs"
 env_logger = "0.5"
 image = "0.19"
 log = "0.4"
-winit = "0.17"
+#winit = "0.17"
```
Comment (ZeGentzy) [inline examples/Cargo.toml]: Should be undone when finished.

**Comment by ZeGentzy** [comment]
The goal was to run gfx-portability ontop of the opengl backend, and I think it's been a partial success. I think this PR will allow us to do that, with the one caveat that it must be OpenGL 3.x+ (4.x?), and X11 only.

Honestly, I don't think it will ever be worth merging, but I'm sure it will be useful neverless.

File: src/backend/gl/src/lib.rs [hunk 14]
```diff
@@ -197,6 +334,18 @@ impl PhysicalDevice {
         }
         let name = info.platform_name.renderer.into();
 
+        // Because some drivers don't actually support this stuff, but report
+        // that they do, we got to disable their private cap.
+        {
+            if !gl.GenVertexArrays.is_loaded {
```
Comment (brendanzab) [inline src/backend/gl/src/lib.rs]: Can you call `gl.GenVertexArrays.is_loaded()` here instead?

Comment (ZeGentzy) [reply src/backend/gl/src/lib.rs]: Haha, I didn't actually see that function when scrolling through the generated file. Thanks.

Comment (brendanzab) [reply src/backend/gl/src/lib.rs]: Yeah, might be more of a bug with my documentation, if anything!

File: src/backend/gl/src/window/glutin.rs [hunk 15]
```diff
@@ -43,65 +43,140 @@
...
+
+pub struct SwapchainImage {
+    // The fbo on the window's context
+    pub(crate) fbo: u32,
+
+    // The image we render too.
+    pub(crate) image: native::Image,
+    pub(crate) image_mem: native::Memory,
+
+    pub(crate) currently_acquired: Mutex<bool>,
```
[the remark this answers is not in the document]
Comment (ZeGentzy) [reply src/backend/gl/src/window/glutin.rs]: `present` only gets an immutable reference of the swapchain.

File: src/backend/gl/src/window/glutin.rs [hunk 16]
```diff
@@ -43,65 +43,140 @@
...
+    pub(crate) extent: hal::window::Extent2D,
+}
+
+pub struct SwapchainImage {
+    // The fbo on the window's context
+    pub(crate) fbo: u32,
+
+    // The image we render too.
+    pub(crate) image: native::Image,
+    pub(crate) image_mem: native::Memory,
```
[the remark this answers is not in the document]
Comment (ZeGentzy) [reply src/backend/gl/src/window/glutin.rs]: Don't we need to deallocate `image_mem` after we are done?

File: src/backend/gl/src/window/glutin.rs [hunk 17]
```diff
@@ -43,65 +43,140 @@
...
-    }
+pub enum Surface {
+    /// This surface has never had a swapchain made on it, and therefore is
+    /// incomplete.
+    PreSwapchained(Arc<SurfacePreSwapchained>),
 
-    fn swapchain_formats(&self) -> Vec<f::Format> {
-        let pixel_format = self.window.get_pixel_format();
+    /// This surface currently has or once had a swapchain made on it, and
+    /// therefore is complete.
```
[the remark this answers is not in the document]
Comment (ZeGentzy) [reply src/backend/gl/src/window/glutin.rs]: Yes, but we would have to dispose of the context and make a new one.

File: src/backend/gl/src/window/glutin.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (msiglreith) [reply src/backend/gl/src/window/glutin.rs]: GL has no notion of memory. No need to jump trough hoops and just manually assemble the image object IMO.

File: src/backend/gl/src/window/glutin.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (msiglreith) [reply src/backend/gl/src/window/glutin.rs]: hmm I see

File: src/backend/gl/src/window/glutin.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (ZeGentzy) [reply src/backend/gl/src/window/glutin.rs]: Will do. :+1:

File: src/backend/gl/src/window/glutin.rs [hunk 18]
```diff
@@ -43,65 +43,140 @@
...
+    pub(crate) device: Device,
+}
 
-    pub fn window(&self) -> &glutin::GlWindow {
-        &self.window
-    }
+pub enum Surface {
+    /// This surface has never had a swapchain made on it, and therefore is
+    /// incomplete.
+    PreSwapchained(Arc<SurfacePreSwapchained>),
```
[the remark this answers is not in the document]
Comment (ZeGentzy) [reply src/backend/gl/src/window/glutin.rs]: One of them was so we could share something, and the other was so we could overwrite surface after cloning the arc. Can't remember which was which,

No longer relevant, as of the newer version.

**Comment by Thomspoon** [comment]
What's the status on this? I saw a bunch of changes in glutin to incorporate these changes, are there still any major hiccups?

**Comment by ZeGentzy** [comment]
@Thomspoon When I get around to updating this, it will be possible to run gfx-portability on top of gl backend on X11 only.

Afraid it does not work with Wayland in most cases because of this:
> One notable limitation of the Wayland backend when it comes to shared contexts is that both contexts must use the same events loop.

Resolving this limitation would require a patch to mesa which I may or may not write some time Soon (tm)

**Comment by Thomspoon** [comment]
@ZeGentzy Would that mean that the opengl backend would not be able to be used in a wasm/emscripten context? I'm not familiar with the benefits/drawbacks of X11 over Wayland (if there are any).

**Comment by ZeGentzy** [comment]
@Thomspoon This PR makes no attempt at supporting wasm/asmjs.

**Comment by kvark** [comment]
@ZeGentzy what's the status of this work?

**Comment by ZeGentzy** [comment]
I'll get around to updating someday, but that day isn't today.

**Comment by kvark** [comment]
@ZeGentzy ok, thank you!
I wasn't meaning to close it. It's totally fine to keep a PR open if you are not abandoning it.

**Comment by ZeGentzy** [comment]
Don't worry, @kvark, any moment now you'll have a attempt #3 on your desk, which if I recall correctly won't require the cross-context bilt. It just needs v0.23.0 of glutin out the door first.
